src/server/game/Spells/Spell.cpp

```diff
@@ -1283,6 +1283,181 @@ void Spell::SelectImplicitCasterDestTargets(SpellEffIndex effIndex, SpellImplici
+                        tstZ = map->GetHeight(phasemask, tstX, tstY, prevZ + travelDistZ, true);
+
+                        if (!map->IsInWater(x, y, z))
+                        {
+                            if (map->IsInWater(tstX, tstY, tstZ) && !map->IsInWater(prevX, prevY, prevZ))// if first we start contact with water, we save coordinate Z before water and use her
+                            {
+                                beforewaterz = prevZ;
+                                tstZ = beforewaterz;
+                            }
+                            else if (map->IsInWater(tstX, tstY, tstZ)) // it next step , where first contact was previos step, and we must recalculate prevZ to Z before water.
+                            {
+                                prevZ = beforewaterz;
+                                tstZ = beforewaterz;
+                            }
+                        }
+                        else if (map->IsInWater(tstX, tstY, tstZ))
+                        {
+                            prevZ = z;
+                            tstZ = z;
+                        }
+
+                        if (!map->IsInWater(tstX, tstY, tstZ))  // second safety check z for blink way if on the ground
+                        {
+                            // highest available point
+                            tstZ1 = map->GetHeight(phasemask, tstX, tstY, prevZ + travelDistZ + 2.0f);
+                            // upper or floor
+                            tstZ2 = map->GetHeight(phasemask, tstX, tstY, prevZ + travelDistZ);
+                            //lower than floor
+                            tstZ3 = map->GetHeight(phasemask, tstX, tstY, prevZ - travelDistZ);
+
+                            //distance of rays, will select the shortest in 3D
+                            srange1 = sqrt((tstY - prevY)*(tstY - prevY) + (tstX - prevX)*(tstX - prevX) + (tstZ1 - prevZ)*(tstZ1 - prevZ));
+                            srange2 = sqrt((tstY - prevY)*(tstY - prevY) + (tstX - prevX)*(tstX - prevX) + (tstZ2 - prevZ)*(tstZ2 - prevZ));
+                            srange3 = sqrt((tstY - prevY)*(tstY - prevY) + (tstX - prevX)*(tstX - prevX) + (tstZ3 - prevZ)*(tstZ3 - prevZ));
+
+                            if (srange1 < srange2)
+                                tstZ = tstZ1 + 0.5f;
+                            else if (srange3 < srange2)
+                                tstZ = tstZ3 + 0.5f;
+                            else
+                                tstZ = tstZ2 + 0.5f;
+                        }
+
+                        destz = tstZ;
+                        srange = sqrt((tstY - prevY)*(tstY - prevY) + (tstX - prevX)*(tstX - prevX) + (tstZ - prevZ)*(tstZ - prevZ));
+                        totalpath += srange;
+                        //TC_LOG_ERROR("server", "Blink cycle checking coordinates, number of cycle = %i , distance of step = %f, total path = %f", j, srange, totalpath);
```

Comment:
Nope. If user will use DEBUG log in config by standart - server will bring lags each blink. Only by error log config available without lag- it first. And second - this needed only for investigation of blink in specific situations.